### ISSUE TITLE
Improved support for simultaneous fitting

### ISSUE THREAD

**Issue by watsonjj** [scikit-hep/iminuit#621]
A nice feature of probfit is the `SimultaneousFit` class (https://probfit.readthedocs.io/en/latest/api.html#simultaneous-fit). 

It appears that iminuit itself is already mostly capable of the features that `SimultaneousFit` provides (through its `CostSum` class). However, unless I am mistaken, there are a few convenience methods missing that would complete this feature. As probfit is deprecated, I would like to contribute this additional functionality to iminuit.

My motivation for this is my single-photoelectron spectra fitting package (https://gitlab.cta-observatory.org/cta-consortium/aswg/tools/spefit), which already utilises iminuit. I was previously unaware of these existing simultaneous-fit features when originally writing this package. I would like to release a new version of `spefit` that replaces much of its own simultaneous-fit bookkeeping with that already provided by iminuit.

Here I demonstrate a version of the probfit `SimultaneousFit` example (https://probfit.readthedocs.io/en/latest/pyplots/costfunc/simul.py) which solely utilises iminuit to achieve the same result:

```python
from iminuit import Minuit
from iminuit.cost import UnbinnedNLL
from iminuit.util import make_func_code
from matplotlib import pyplot as plt
import numpy as np
from numpy.random import randn, seed
from numba import vectorize, float64
from math import exp, sqrt, pi

SQRT2PI = sqrt(2.0 * pi)

@vectorize([float64(float64, float64, float64)], fastmath=True)
def normal_pdf(x, mean, std_deviation):
    u = (x - mean) / std_deviation
    return exp(-0.5 * u ** 2) / (SQRT2PI * std_deviation)

seed(0)
width = 2.0
data1 = randn(1000) * width + 1
data2 = randn(1000) * width + 2

normal_pdf.func_code = make_func_code(("x", "mu_1", "sigma"))
lh1 = UnbinnedNLL(data1, normal_pdf)
normal_pdf.func_code = make_func_code(("x", "mu_2", "sigma")) # TODO: Add method to rename lh1
lh2 = UnbinnedNLL(data2, normal_pdf)

simlh = lh1 + lh2

m = Minuit(simlh, mu_1=1.2, mu_2=2.2, sigma=1.5)

def plot(cost, minuit, i, ax):
    names_i = [minuit.parameters[map_i] for map_i in simlh._maps[i]]  # TODO: Add method to BasicView to obtain view subset
    values_i = [minuit.values[map_i] for map_i in simlh._maps[i]]  # TODO: Add method to BasicView to obtain view subset
    errors_i = [minuit.errors[map_i] for map_i in simlh._maps[i]]  # TODO: Add method to BasicView to obtain view subset
    data_i = cost[i].data
    model_i = cost[i]._model   # TODO: Expose public access in Cost? Unfortunately not common between Cost methods...
    bins, edges = np.histogram(data_i, bins=100)
    mid = (edges[1:] + edges[:-1]) / 2
    binwidth = (edges[1] - edges[0])
    ax.errorbar(mid, bins, np.sqrt(bins), fmt="b.", capsize=0, zorder=0)
    y = model_i(mid, *values_i) * np.sum(bins * binwidth)
    ax.plot(mid, y)
    t = ""
    for n, v, e in zip(names_i, values_i, errors_i):
        t += f"{n} = {v:.3f} ± {e:.3f}\n"
    ax.text(0.01, 0.99, t, transform=ax.transAxes, va='top')
    
fig = plt.figure(figsize=(14, 5))
ax1 = fig.add_subplot(1, 2, 1)
ax2 = fig.add_subplot(1, 2, 2)
plot(simlh, m, 0, ax1)
plot(simlh, m, 1, ax2)
fig.suptitle("Before")

m.migrad()  # fit

fig = plt.figure(figsize=(14, 5))
ax1 = fig.add_subplot(1, 2, 1)
ax2 = fig.add_subplot(1, 2, 2)
plot(simlh, m, 0, ax1)
plot(simlh, m, 1, ax2)
fig.suptitle("After")
```
![image](https://user-images.githubusercontent.com/17825673/111623627-e50f1500-87ea-11eb-89ea-f132cef06080.png)

The TODOs highlight the areas where functionality might be added to iminuit to better support this feature. Please correct me if I have overlooked some existing features in iminuit which already provide this.

To summarise, the main features which appear to be missing are:
- Nice way to rename the func code to indicate which parameters are shared vs unique
- Access a subset of the ValueView for each individual fit.

**Comment by mayou36** [comment]
Hi, in general the idea is to keep the iminuit package as a minimizer package and to put extended fitting capabilities and losses into other packages. It the unbinned case, I would recommend to have a look at [zfit](https://github.com/zfit/zfit) which can take care of simultaneous fits and shared parameters and is "the replacement" for probfit. If anything is missing, please feel free to make an issue about this.

**Comment by watsonjj** [comment]
I have looked at zfit as well. While it is really a great package, it goes beyond my needs and yet doesn't include binned fitting (which I need for my package). I see there is already an issue open for that. 

As demonstrated above, `iminuit` already has all the capabilities to support simultaneous fitting. I believe the two inclusions I am suggesting are minor extensions to the existing functionality. I plan to open a PR soon with my implementation of these features. I created this issue to check if I was overlooking some parts of iminuit that already implements the convenient methods I was searching for. I could not find them, nor could I find an example of utilising the `CostSum` for simultaneous fitting in this way...is there a reason for that?

**Comment by mayou36** [comment]
> it goes beyond my needs

That may be true, however, being a simple case in a more powerful package has it's benefit of being well covered. zfit is meant to be a powerful "base-package" where other libraries can build on top. This includes a nice API and the possibility for enough customizations.

> and yet doesn't include binned fitting (which I need for my package). I see there is already an issue open for that.

Indeed! It seems tricky to get that right. However, it is by far flexible enough to do binned fits (5-10 lines of code), some people do it already.

Looking at the snippet, there seems to be a lot of resemblance with e.g. the basic example for binned fits in zfit (this is unmodified and not tailored to anything):

```
obs = zfit.Space("x", limits=(-10, 10))

# parameters
mu_shared = zfit.Parameter("mu_shared", 1., -4, 6)  # mu is a shared parameter
sigma1 = zfit.Parameter("sigma_one", 1., 0.1, 10)
sigma2 = zfit.Parameter("sigma_two", 1., 0.1, 10)

# model building, pdf creation
gauss1 = zfit.pdf.Gauss(mu=mu_shared, sigma=sigma1, obs=obs)
gauss2 = zfit.pdf.Gauss(mu=mu_shared, sigma=sigma2, obs=obs)

# data
normal_np = np.random.normal(loc=2., scale=3., size=10000)
data1 = zfit.Data.from_numpy(obs=obs, array=normal_np)

# data
normal_np = np.random.normal(loc=2., scale=4., size=10000)
data2 = zfit.Data.from_numpy(obs=obs, array=normal_np)

# create simultaenous loss, two possibilities
nll_simultaneous = zfit.loss.UnbinnedNLL(model=[gauss1, gauss2],
                                         data=[data1, data2])
# OR, equivalently
nll1 = zfit.loss.UnbinnedNLL(model=gauss1, data=data1)
nll2 = zfit.loss.UnbinnedNLL(model=gauss2, data=data2)
nll_simultaneous2 = nll1 + nll2

minimizer = zfit.minimize.Minuit()
minimizer.minimize(nll_simultaneous2)
```

The advantage is that all the shared parameters etc are all well defined.

In fact, having looked through the package it may be worth to have a chat, I guess there could be quite some things to be shared (you can also contact me by mail Jonas.Eschle AT cern.ch)

And in case there is nothing to collaborate on, maybe zfit doesn't suit at all, I'd be very glad anyway to better understand the library and get an idea of the package as a use-case.

**Comment by watsonjj** [comment]
> In fact, having looked through the package it may be worth to have a chat, I guess there could be quite some things to be shared (you can also contact me by mail Jonas.Eschle AT cern.ch)

This sounds great - I will contact you.

>Looking at the snippet, there seems to be a lot of resemblance with e.g. the basic example for binned fits in zfit (this is unmodified and not tailored to anything)

Thank you for the example using zfit. However, this still appears to still be an unbinned fit (just as my example also was). If instead of the `normal_np` array, the user only had access to the result of `bins, edges = np.histogram(normal_np, bins=100)`, is this case handled by zfits at all?

**Comment by HDembinski** [comment]
@watsonjj Note that you can get numba-accelerated functions like normal_pdf from my other package [numba-stats](https://github.com/HDembinski/numba-stats)

I think your suggestions make sense, we need to open some specialized issues for them.

**Comment by HDembinski** [comment]
> Hi, in general the idea is to keep the iminuit package as a minimizer package and to put extended fitting capabilities and losses into other packages. It the unbinned case, I would recommend to have a look at [zfit](https://github.com/zfit/zfit) which can take care of simultaneous fits and shared parameters and is "the replacement" for probfit. If anything is missing, please feel free to make an issue about this.

This is not correct. iminuit will include commonly used cost functions and offer ways to combine them. What we won't support is a high-level interface build models.

**Comment by mayou36** [comment]
> This is not correct. iminuit will include commonly used cost functions and offer ways to combine them. What we won't support is a high-level interface to build models, for that you should use for example zfit.

Thanks for the clarification, I was not quite aware of the full scope of iminuit, apologies!

**Comment by HDembinski** [comment]
Closing this since the improvements were merged.
